File: src/lib/bpm-calculator.ts
/**
 * Dutch BPM (Belasting van Personenauto's en Motorrijwielen) Calculator
 * Based on 2026 rates from Belastingdienst
 * Uses Forfaitaire Tabel method for depreciation
 */

// 2026 BPM CO2 Tariff Brackets
const BPM_BRACKETS_2026 = [
  { minCO2: 0, maxCO2: 79, ratePerGram: 0, baseAmount: 667 },
  { minCO2: 80, maxCO2: 124, ratePerGram: 6.68, baseAmount: 0 },
  { minCO2: 125, maxCO2: 169, ratePerGram: 67.40, baseAmount: 0 },
  { minCO2: 170, maxCO2: 199, ratePerGram: 159.61, baseAmount: 0 },
  { minCO2: 200, maxCO2: Infinity, ratePerGram: 490.91, baseAmount: 0 },
];

// Diesel surcharge: €109.87 per g/km above 70 g/km
const DIESEL_SURCHARGE_THRESHOLD = 70;
const DIESEL_SURCHARGE_RATE = 109.87;

// Forfaitaire Depreciation Table (2026)
// Based on vehicle age in months
const DEPRECIATION_TABLE: { maxMonths: number; percentage: number }[] = [
  { maxMonths: 3, percentage: 0 },
  { maxMonths: 6, percentage: 24 },
  { maxMonths: 9, percentage: 33 },
  { maxMonths: 18, percentage: 42 },
  { maxMonths: 24, percentage: 49 },
  { maxMonths: 36, percentage: 56 },
  { maxMonths: 48, percentage: 63 },
  { maxMonths: 60, percentage: 70 },
  { maxMonths: 72, percentage: 76 },
  { maxMonths: 84, percentage: 81 },
  { maxMonths: 96, percentage: 85 },
  { maxMonths: 108, percentage: 88 },
  { maxMonths: 120, percentage: 90 },
  { maxMonths: Infinity, percentage: 92 },
];

export type FuelType = 'petrol' | 'diesel' | 'electric' | 'hybrid' | 'lpg';

export interface BPMInput {
  co2_gkm: number; // CO2 emissions in g/km (WLTP)
  fuelType: FuelType;
  firstRegistrationDate: Date; // First registration date (abroad)
  calculationDate?: Date; // Date of BPM calculation (defaults to now)
}

export interface BPMResult {
  grossBPM: number; // BPM before depreciation
  dieselSurcharge: number; // Additional diesel tax
  totalGrossBPM: number; // Gross + diesel surcharge
  vehicleAgeMonths: number;
  depreciationPercentage: number;
  restBPM: number; // Final BPM to pay
  breakdown: {
    baseBPM: number;
    co2BPM: number;
    dieselSurcharge: number;
  };
}

/**
 * Calculate the number of months between two dates
 */
function monthsBetween(startDate: Date, endDate: Date): number {
  const years = endDate.getFullYear() - startDate.getFullYear();
  const months = endDate.getMonth() - startDate.getMonth();
  return years * 12 + months;
}

/**
 * Get depreciation percentage from forfaitaire table based on vehicle age
 */
function getDepreciationPercentage(ageMonths: number): number {
  for (const bracket of DEPRECIATION_TABLE) {
    if (ageMonths <= bracket.maxMonths) {
      return bracket.percentage;
    }
  }
  return 92; // Maximum depreciation
}

/**
 * Calculate gross BPM based on CO2 emissions (before depreciation)
 */
function calculateGrossBPM(co2_gkm: number): { baseBPM: number; co2BPM: number; total: number } {
  // Base amount (fixed rate for 2026)
  const baseBPM = 667;

  // Electric vehicles (0 CO2) only pay base rate
  if (co2_gkm === 0) {
    return { baseBPM, co2BPM: 0, total: baseBPM };
  }

  let co2BPM = 0;

  // Calculate CO2-based BPM using progressive brackets
  // Bracket 1: 0-79 g/km = €0 (only base rate)
  // Bracket 2: 80-124 g/km = €6.68 per gram
  if (co2_gkm > 79) {
    const gramsInBracket = Math.min(co2_gkm, 124) - 79;
    co2BPM += gramsInBracket * 6.68;
  }

  // Bracket 3: 125-169 g/km = €67.40 per gram
  if (co2_gkm > 124) {
    const gramsInBracket = Math.min(co2_gkm, 169) - 124;
    co2BPM += gramsInBracket * 67.40;
  }

  // Bracket 4: 170-199 g/km = €159.61 per gram
  if (co2_gkm > 169) {
    const gramsInBracket = Math.min(co2_gkm, 199) - 169;
    co2BPM += gramsInBracket * 159.61;
  }

  // Bracket 5: 200+ g/km = €490.91 per gram
  if (co2_gkm > 199) {
    const gramsInBracket = co2_gkm - 199;
    co2BPM += gramsInBracket * 490.91;
  }

  return {
    baseBPM,
    co2BPM: Math.round(co2BPM * 100) / 100,
    total: Math.round((baseBPM + co2BPM) * 100) / 100,
  };
}

/**
 * Calculate diesel surcharge
 * Applies when CO2 > 70 g/km for diesel vehicles
 */
function calculateDieselSurcharge(co2_gkm: number, fuelType: FuelType): number {
  if (fuelType !== 'diesel') {
    return 0;
  }

  if (co2_gkm <= DIESEL_SURCHARGE_THRESHOLD) {
    return 0;
  }

  const gramsAboveThreshold = co2_gkm - DIESEL_SURCHARGE_THRESHOLD;
  return Math.round(gramsAboveThreshold * DIESEL_SURCHARGE_RATE * 100) / 100;
}

/**
 * Main BPM calculation function
 * Implements the Forfaitaire Tabel method
 */
export function calculateBPM(input: BPMInput): BPMResult {
  const calculationDate = input.calculationDate || new Date();

  // Calculate vehicle age in months
  const vehicleAgeMonths = monthsBetween(input.firstRegistrationDate, calculationDate);

  // Get depreciation percentage from forfaitaire table
  const depreciationPercentage = getDepreciationPercentage(vehicleAgeMonths);

  // Calculate gross BPM components
  const grossBPMComponents = calculateGrossBPM(input.co2_gkm);

  // Calculate diesel surcharge
  const dieselSurcharge = calculateDieselSurcharge(input.co2_gkm, input.fuelType);

  // Total gross BPM (before depreciation)
  const totalGrossBPM = grossBPMComponents.total + dieselSurcharge;

  // Apply depreciation to get Rest-BPM
  const depreciationFactor = 1 - depreciationPercentage / 100;
  const restBPM = Math.round(totalGrossBPM * depreciationFactor * 100) / 100;

  return {
    grossBPM: grossBPMComponents.total,
    dieselSurcharge,
    totalGrossBPM,
    vehicleAgeMonths,
    depreciationPercentage,
    restBPM,
    breakdown: {
      baseBPM: grossBPMComponents.baseBPM,
      co2BPM: grossBPMComponents.co2BPM,
      dieselSurcharge,
    },
  };
}

/**
 * Quick calculation helper for common use case
 */
export function quickBPMEstimate(
  co2_gkm: number,
  fuelType: FuelType,
  vehicleAgeMonths: number
): number {
  // Create a synthetic first registration date based on age
  const now = new Date();
  const firstRegistrationDate = new Date(now);
  firstRegistrationDate.setMonth(now.getMonth() - vehicleAgeMonths);

  const result = calculateBPM({
    co2_gkm,
    fuelType,
    firstRegistrationDate,
    calculationDate: now,
  });

  return result.restBPM;
}

// For use in n8n Code nodes (CommonJS export)
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { calculateBPM, quickBPMEstimate };
}

File: modal-app/bpm_calculator.py
"""
BPM Calculator for Dutch vehicle import tax.
Based on 2026 Belastingdienst forfaitaire tabel.
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Optional

from constants import CO2_BRACKETS, DIESEL_SURCHARGE_RATE, DIESEL_SURCHARGE_THRESHOLD, DIESEL_SURCHARGE_SUBTRACT, DEPRECIATION_TABLE
from utils import calculate_vehicle_age_months, normalize_fuel_type


@dataclass
class BPMResult:
    """Result of BPM calculation."""
    gross_bpm: float
    rest_bpm: float
    depreciation_percentage: float
    diesel_surcharge: float
    vehicle_age_months: int
    co2_gkm: int
    fuel_type: str


def calculate_gross_bpm(co2_gkm: int) -> float:
    """
    Calculate gross BPM based on CO2 emissions.
    Official formula: BPM = (CO2 - threshold) * rate + base
    Where threshold is "kolom 1" from the official Belastingdienst table.

    Args:
        co2_gkm: CO2 emissions in g/km

    Returns:
        Gross BPM amount in euros
    """
    # Find the correct bracket for this CO2 value
    for bracket in CO2_BRACKETS:
        if bracket["min"] <= co2_gkm <= bracket["max"]:
            # Formula: (CO2 - threshold) * rate + base
            # threshold is the official "kolom 1" value from Belastingdienst
            bpm = (co2_gkm - bracket["threshold"]) * bracket["rate"] + bracket["base"]
            return round(bpm, 2)

    # If beyond all brackets, use the last bracket
    last_bracket = CO2_BRACKETS[-1]
    bpm = (co2_gkm - last_bracket["threshold"]) * last_bracket["rate"] + last_bracket["base"]
    return round(bpm, 2)


def calculate_diesel_surcharge(co2_gkm: int, fuel_type: str) -> float:
    """
    Calculate diesel surcharge if applicable.
    Official formula: (CO2 - 69) × €114,83 for diesel cars with CO2 > 70 g/km

    Args:
        co2_gkm: CO2 emissions in g/km
        fuel_type: Normalized fuel type

    Returns:
        Diesel surcharge amount in euros
    """
    if fuel_type != "diesel":
        return 0

    if co2_gkm <= DIESEL_SURCHARGE_THRESHOLD:
        return 0

    # Official formula subtracts 69, not the threshold (70)
    excess_grams = co2_gkm - DIESEL_SURCHARGE_SUBTRACT
    return round(excess_grams * DIESEL_SURCHARGE_RATE, 2)


def get_depreciation_percentage(age_months: int) -> float:
    """
    Get depreciation percentage from forfaitaire tabel.
    Uses official method: base_percentage + (months_in_period * monthly_addition)

    Args:
        age_months: Vehicle age in months

    Returns:
        Depreciation percentage (0-100)
    """
    for bracket in DEPRECIATION_TABLE:
        if bracket["min_months"] <= age_months <= bracket["max_months"]:
            # Calculate months within this bracket
            months_in_bracket = age_months - bracket["min_months"]

            # Calculate depreciation: base + (months * monthly_addition)
            depreciation = bracket["base_percentage"] + (months_in_bracket * bracket["monthly_addition"])

            # Cap at 100%
            return min(depreciation, 100.0)

    # If beyond all brackets, return maximum from last bracket
    last_bracket = DEPRECIATION_TABLE[-1]
    return min(last_bracket["base_percentage"] + (12 * last_bracket["monthly_addition"]), 100.0)


def calculate_bpm(
    co2_gkm: int,
    fuel_type: str,
    first_registration_date: datetime
) -> BPMResult:
    """
    Calculate complete BPM for a vehicle.

    Args:
        co2_gkm: CO2 emissions in g/km
        fuel_type: Fuel type string
        first_registration_date: Date of first registration

    Returns:
        BPMResult with all calculation details
    """
    # Normalize fuel type
    normalized_fuel = normalize_fuel_type(fuel_type)

    # Calculate vehicle age
    age_months = calculate_vehicle_age_months(first_registration_date)

    # Calculate gross BPM
    gross_bpm = calculate_gross_bpm(co2_gkm)

    # Calculate diesel surcharge
    diesel_surcharge = calculate_diesel_surcharge(co2_gkm, normalized_fuel)

    # Total gross BPM including surcharge
    total_gross = gross_bpm + diesel_surcharge

    # Get depreciation percentage
    depreciation = get_depreciation_percentage(age_months)

    # Calculate rest BPM (what you actually pay)
    rest_bpm = total_gross * (1 - depreciation / 100)

    return BPMResult(
        gross_bpm=round(gross_bpm, 2),
        rest_bpm=round(rest_bpm, 2),
        depreciation_percentage=depreciation,
        diesel_surcharge=round(diesel_surcharge, 2),
        vehicle_age_months=age_months,
        co2_gkm=co2_gkm,
        fuel_type=normalized_fuel,
    )


def bpm_to_dict(result: BPMResult) -> dict:
    """Convert BPMResult to dictionary for JSON serialization."""
    return {
        "grossBPM": result.gross_bpm,
        "restBPM": result.rest_bpm,
        "depreciationPercentage": result.depreciation_percentage,
        "dieselSurcharge": result.diesel_surcharge,
        "vehicleAgeMonths": result.vehicle_age_months,
        "co2_gkm": result.co2_gkm,
        "fuelType": result.fuel_type,
    }
